### PR TITLE
Emit source positions for open/close braces

### LINKED ISSUE #39170: sourcemap mapping is generated before node generates newline and indent
**TypeScript Version:** 4.0.0-dev.20200615

**Search Terms:**
sourcemap

**Code**

```ts
export type AsyncFn<TResult> = () => PromiseLike<TResult>;

export type RaceCancelFn = <TResult>(
  asyncFnOrPromise: AsyncFn<TResult> | PromiseLike<TResult>
) => Promise<TResult>;

export type DisposeFn = () => void;

export type DisposableExecutorFn<TResult> = (
  resolve: (value?: TResult | PromiseLike<TResult>) => void,
  reject: (reason?: any) => void
) => DisposeFn;

const noopRaceCancel: RaceCancelFn = <TResult>(
  asyncFnOrPromise: AsyncFn<TResult> | PromiseLike<TResult>
): Promise<TResult> =>
  typeof asyncFnOrPromise === "function"
    ? Promise.resolve().then(asyncFnOrPromise)
    : Promise.resolve(asyncFnOrPromise);

export default async function cancellablePromise<TResult>(
  disposableExecutor: DisposableExecutorFn<TResult>,
  raceCancel: RaceCancelFn = noopRaceCancel
): Promise<TResult> {
  let dispose: DisposeFn | undefined;
  try {
    return await raceCancel(
      () =>
        new Promise<TResult>((resolve, reject) => {
          dispose = disposableExecutor(resolve, reject);
        })
    );
  } finally {
    if (dispose !== undefined) {
      dispose();
    }
  }
}
```
copy above code to index.ts then run:
`tsc --target esnext --sourceMap index.ts`

**Expected behavior:**
I expect that since this should just be removing types that the generated sourcemap could map all JS AST nodes start/end back to the original but at a minimum starts of statements, expressions, identifiers, blocks.  Ideally for source map generation identifiers would 1 to 1 map to identifiers and not include type.  Ideally starts should not include whitespace.

**Actual behavior:**
Missing closing braces in the try/finally block and the if in the finally block (since tsc 2.8).
Closing parens often start on prior lines ending newline instead of the paren in the original (since as far back as tsc 2.0).
Identifiers missing mapping at end when typing info was removed or there are default expressions.

The last one maybe considered a nice to have but is helpful with various tooling and to survive downstream transforms like rollup and minification. 

tsc-sourcemap

### PR DESCRIPTION
**Pull request by rbuckton**
This changes the emitter to emit source positions before and after open and close braces (`{`, `}`) so that tools like `v8-to-istanbul` can correctly map branches for code coverage. Brace source positions are ignored when the output is a declaration file.

NOTE: This does not emit source positions for *all* tokens that pass through `emitTokenWithComment`, as that significantly increases the size of the output `.js.map` file.

Fixes #39170

### CODE REVIEW

**Comment by rbuckton** [comment]
@typescript-bot perf test
@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -455,6 +455,7 @@ namespace ts {
                 inlineSourceMap: compilerOptions.inlineSourceMap,
                 extendedDiagnostics: compilerOptions.extendedDiagnostics,
                 onlyPrintJsDocStyle: true,
+                omitBraceSourcePositions: true,
```
Comment (andrewbranch) [inline src/compiler/emitter.ts:458]: Can we make this name more obviously about source _maps_? I think most people dealing with the emitter API have .js/.d.ts emit in their minds, where this name makes no sense.

Comment (rbuckton) [reply src/compiler/emitter.ts:458]: I can change the name, though this property is marked `/*@internal*/` so it shouldn't impact API consumers outside of our own team.

**Review by sandersn — APPROVED** [review]

**Review by andrewbranch — APPROVED** [review]

**Comment by sandersn** [comment]
@rbuckton I'm going through all the ready-to-merge PRs, and this one is all signed off. Is it worth bringing up to date and merging it for 5.1?

**Comment by sandersn** [comment]
@rbuckton Same question as before, but for 5.3 -- should we take this one or close it?

**Comment by rbuckton** [comment]
I'll see if I can get it up to date.

File: tests/cases/fourslash/server/declarationMapsEnableMapping_NoInline.ts [hunk 2]
```diff
@@ -1,3 +1,89 @@
+<<<<<<< HEAD
```
Comment (rbuckton) [inline tests/cases/fourslash/server/declarationMapsEnableMapping_NoInline.ts:1]: VSCode did not detect this as a merge conflict for some reason. I'll post a fix shortly.

**Pull request merged**